### DISCUSSION TITLE
Enhance PAPERLESS_FILENAME_FORMAT with structure through tag selection/class or slashes

### DISCUSSION THREAD

**Discussion by lenucksi** [jonaswinkler/paperless-ng#803] [Feature Requests]
I'm using paperless to convert my stash of scanned and non-scanned documents into a searchable, sorted and consistently named stash of documents that I later access via SMB to process the documents with other applications.

Also, I'm interested in always keeping a file tree that has structure and is usable entirely without paperless (a second effect of the first goal listed above)

I used to have a tree structure like `year/{bank|telco|work}/{vendorname}/[files]`

To create something this I'm using `PAPERLESS_FILENAME_FORMAT` and a matching string. I tried to use document types with slashes such as `bill/telecommunications` but found them to be rendered as `bill-telecommunications` after slash replacement. Using document types for such structure efforts also didn't seen right.
I found the opportunity to use the tags (which I'm assigning and generating from a set of ingestion folders) in the filename configuration. However, I found that I can only dump all tags assigned to a document and not structure, select or use them in other ways to create effect described above.
I don't have a ready-to-code idea of how to do such selection but I thought I'd write this up as someone else might have ideas how to do such a selection/structuring or already came up with similar ways to do this.

**Comment by benjaminfrank** [comment]
I think [here](https://github.com/jonaswinkler/paperless-ng/blob/master/src/documents/file_handling.py#L125) is the right place where this could be done.
PAPERLESS_FILENAME_FORMAT` seems to be utilizing the standard [`format`](https://docs.python.org/3.4/library/string.html#string-formatting). 

One way I could imagine here is to specify a prefix and use the tag(s) that satisfy the prefix:
with `PAPERLESS_FILENAME_FORMAT = {year}/{tag_prefix("foo:")}/{vendorname}` 
a file having a tag `foo:me` the path would be something like `2021/me/USPO` 
a file with tags `foo:me foo:you foo:them` would be `2021/me-you-them/USPO`

implementing `tag_prefix("foo:")` should not be that hard but wiring it into `format` might be a bit trickier
